### ISSUE TITLE
Environments Entry for "meteor"

### ISSUE THREAD

**Issue by rhettlivingston** [Galooshi/import-js#289]
I second @awatson1978 in saying "thank you, thank you for putting this together"!

To complement #288, I am investigating other aspects of getting this package to play really nice with Meteor.

Though node.js is a part of meteor under the hood, what we see as our "core" modules is quite different. As a start to get the groupImports feature to play nicely with Meteor, I'd like to request an addition to the supported "environments". I am currently vetting my list of what I believe are Meteor's "core" modules on their forum, but believe it will be close to the following:

```js
  meteor: [
    'meteor/accounts-base',
    'meteor/blaze',
    'meteor/check',
    'meteor/ddp-client',
    'meteor/ddp-rate-limiter',
    'meteor/ejson',
    'meteor/email',
    'meteor/http',
    'meteor/check',
    'meteor/meteor',
    'meteor/mongo',
    'meteor/random',
    'meteor/reactive-var',
    'meteor/session',
    'meteor/templating',
    'meteor/tracker',
  ],
```

Would you be willing to accept a PR for this?

Also, I have several other ideas on the burner for things like 
 - automatically identifying our 3rd party packages (that middle group in groupImports),
 - easing our extensive use of namedExports (unlike NPM, our package system defines the exports so we have the leg up you were discussing in #107 for this one),
 - exposing babel-lint's "allowImportExportEverywhere" option to us (we'll need to turn it on),
 - and handling the common usage of the import for side-effects statement in Meteor (`import './thisModulesName.http;'` and `import './thisModulesName.css';`).

As well as thoughts about the configuration processing (love the idea of accepting a js file and wonder if that could be expanded on a touch to allow plugins that specify things like above for different environments). Should I approach these via more "issues" or a more general offline discussion followed by more issues?

**Comment by trotzig** [comment]
Hi @rhettlivingston, and thanks for the ❤️!

Let's keep the discussion here to begin with, and then we can split it out into separate issues later on. I'll try to give a few thoughts on your initial post. 

> Would you be willing to accept a PR for this? (meteor core modules)

Yes, absolutely.

> automatically identifying our 3rd party packages (that middle group in groupImports),

How are these dependencies defined? (I'm not a Meteor user myself)

> easing our extensive use of namedExports (unlike NPM, our package system defines the exports so we have the leg up you were discussing in #107 for this one),

Great! I think you/we have an opportunity to lead the effort forward here. Perhaps we can get more packages to adopt the Meteor style here. 

> exposing babel-lint's "allowImportExportEverywhere" option to us (we'll need to turn it on),

I'm not sure I understand this. Can you explain a little more in-depth?

> and handling the common usage of the import for side-effects statement in Meteor (import './thisModulesName.http;' and import './thisModulesName.css';).

This is somewhat supported today since you can use [dynamic parts in aliases](https://github.com/Galooshi/import-js#aliases). E.g.
```json
"aliases": {
  "styles": "./{filename}.css",
  "ajax": "./{filename}.http"
}
```
We would just have to make these end up in the config automatically if you use the `meteor` environment.

**Comment by rhettlivingston** [comment]
>> automatically identifying our 3rd party packages (that middle group in groupImports),

>How are these dependencies defined? (I'm not a Meteor user myself)

We run a command, "meteor add packageName", that downloads the package and makes entries in the project's .meteor/packages file. So all of the 3rd party packages will appear there mixed with some core ones. If a package is there and is not a known core package, it is a 3rd party package and should go in the 2nd group of groupImports.

One caveat, in the import statements, it will have "meteor/" prepended to it to indicate that it is a meteor package instead of an NPM package (Meteor supports both).

>>easing our extensive use of namedExports (unlike NPM, our package system defines the exports so we have the leg up you were discussing in #107 for this one),

>Great! I think you/we have an opportunity to lead the effort forward here. Perhaps we can get more packages to adopt the Meteor style here.

Agree. It is very important to us. We have ALOT of namedExports in typical projects. The only difficulty is that this target will probably be a moving one. We are in the midst of a transition from our own package system to NPM. We currently support both. Of course, that may mean that we'll be extending NPM ourselves soon.

This one was nagging me though when thinking about the implementation. It might be more robust for you if your thought about accepting an .importjs.js file for configuration was in place. Then this could be implemented as a function adding to the namedExports. Similarly, options for "coreModules" and "packages" could be supported in your configuration allowing your users to implement these as functions. It would be a minor further step for you to define a "plugin" system that simply supplies these functions from another NPM module. Then we could write a Meteor plugin and other environments could do the same.

But,,, that's perhaps a longer term thought. I don't think it matches my schedule.

>>exposing babel-eslint's "allowImportExportEverywhere" option to us (we'll need to turn it on),

>I'm not sure I understand this. Can you explain a little more in-depth?

We just added support for conditional and nested imports. We needed them to get rid of the need to occasionally use "require" to do something like:

```js
if (Meteor.isClient) {
  const clientonlyFunction = require('./client-only-file.js');
  // code using clientOnlyFunction
  //
}
```

which can now be done by 

```
if (Meteor.isClient) {
  import { clientonlyFunction } from './client-only-file';
  // code using clientOnlyFunction
  //
}
```

We pretty quickly noted that this broke ESLint. It would give a syntax error and stop parsing the file.

Luckily, Henry Zhu, a maintainer of the babel-eslint parser, was quick to accept a PR to add an option, allowImportExportEverywhere, to babel-eslint that just gets passed through to babylon (the underlying parser) which already supported it. So, our .eslintrc.json files just have to have the following in them to fix this issue:

```json
  "parser": "babel-eslint",
  "parserOptions": {
    "allowImportExportEverywhere": true
  },
```

>>and handling the common usage of the import for side-effects statement in Meteor (import './thisModulesName.http;' and import './thisModulesName.css';).

Actually, as these are for side-effect only, there is no indication within the module you're in that they are needed, i.e. no undefined variables. So I don't see how aliases can help.

Basically, we have a common component coding pattern of having a .js, .html, and .css file that together define a UI component. So, if you're processing 'somename.js' and there is a 'somename.html' and/or 'somename.css' file in the same directory, there should be an `import './somename.html';` and/or `import './somename.css';` statements at the top (usually last in all of the imports) of the .js file to accomplish the side-effect of simply getting those files included in the build.

I'm thinking two configuration switches, perhaps `autoImportComponentHTML: true` and `autoImportComponentCSS: true`, could do it. Alternatively, `"autoImportComponent": [ ".html", ".css" ]` might be more robust if meteor adds support for more of these in it's build system (scss, less, etc.).

**Comment by rhettlivingston** [comment]
Just noticed a little conceptual problem with the "meteor" environment while writing the tests.

All of the meteor core modules have named exports. So, the meteor environment is very much incomplete without:

```json
  "namedExports": {
    "meteor/accounts-base": [
      "AccountsClient",
      "Accounts",
      "AccountsServer"
    ],
    "meteor/blaze": [ "Blaze" ],
    "meteor/check": [ "check" ],
    "meteor/ddp-client": [ "DDP" ],
    "meteor/ddp-rate-limiter": [ "DDPRateLimiter" ],
    "meteor/ejson": [ "EJSON" ],
    "meteor/email": [ "Email" ],
    "meteor/http": [ "HTTP" ],
    "meteor/check": [ "check", "Match" ],
    "meteor/meteor": [ "Meteor" ],
    "meteor/mongo": [ "Mongo" ],
    "meteor/random": [ "Random" ],
    "meteor/reactive-var": [ "ReactiveVar" ],
    "meteor/session": [ "Session" ],
    "meteor/templating": [ "Template" ],
    "meteor/tracker": [ "Tracker" ]
  }
```

So, it seems like this should imply the above namedExports. This is similar to your thoughts at Default aliases #232. Interestingly, in my project I implemented underscore and jquery as the following namedExports instead of as aliases.

```json
    "meteor/underscore": [ "_" ],
    "meteor/jquery": [ "$", "jQuery" ],
```

So do you think it would be appropriate to expand the purpose of environment beyond just specifying the core modules so that it also sets appropriate namedExports for those modules? Perhaps other things too like @awatson1978's leading prefix for absolute paths?

Also, I had to set useRelativePaths false to get it to stop appending "./" in front of my core module reference. I think I need to make an adjustment in code recognizing packages somewhere to recognize meteor packages as well as NPM packages.

**Comment by trotzig** [comment]
Thanks for the lengthy reply, I have a pretty good picture of what we need to accomplish now. I think it would make sense to front-load allowing the configuration to be javascript. After that, we can allow environments to specify a full set of configuration options. I'm tracking javascript configuration in https://github.com/Galooshi/import-js/issues/291. 

> One caveat, in the import statements, it will have "meteor/" prepended to it to indicate that it is a meteor package instead of an NPM package (Meteor supports both).

If we allow `aliases` to be a function, we can dynamically construct an alias for each dependency. Or named export, whatever fits best. 

> It would be a minor further step for you to define a "plugin" system that simply supplies these functions from another NPM module. Then we could write a Meteor plugin and other environments could do the same. But,,, that's perhaps a longer term thought. I don't think it matches my schedule.

If we allow js configuration, we are in a way already supporting "plugins", because you could do something like this:

```js
const meteorPresets = require('meteor-importjs-presets);
module.exports = Object.assign(meteorPresets, {
  useRelativePaths: false,
});
```
> We pretty quickly noted that this broke ESLint. It would give a syntax error and stop parsing the file.

Interesting, I recently ran into a similar issue. I think ImportJS could probably also switch over to babel-eslint as well. Although I'd be interested to know if there's a performance difference between the two parsers.

> Actually, as these are for side-effect only, there is no indication within the module you're in that they are needed, i.e. no undefined variables. So I don't see how aliases can help.

Ah, I see. There's actually an open issue for imports that don't assign to variables. So we would have to fix that first if we want to somehow auto-import these. Once we do, I think we could make the configuration option a little more general-purpose by allowing something like this:

```js
function isComponent(file) {
  // do something smart
}

autoImport: function(currentFile) {
  if (isComponent(currentFile)) {
    return [
      path.basename(currentFile) + '.css',
      path.basename(currentFile) + '.html',
    ];
  } 
}
```

> So do you think it would be appropriate to expand the purpose of environment beyond just specifying the core modules so that it also sets appropriate namedExports for those modules? Perhaps other things too like @awatson1978's leading prefix for absolute paths?

Yep. I started tracking that in a separate issue: https://github.com/Galooshi/import-js/issues/290

> Also, I had to set useRelativePaths false to get it to stop appending "./" in front of my core module reference. I think I need to make an adjustment in code recognizing packages somewhere to recognize meteor packages as well as NPM packages.

Ah, that's annoying. I think we will have to add another exception to `Configuration.resolveNamedExports` (before we relativize the path). Here: https://github.com/Galooshi/import-js/blob/a8a3ce020cd650a4993d2bff14ce6e953eea7c3b/lib/Configuration.js#L271

**Comment by trotzig** [comment]
How's this for a plan?

1. Allow javascript configuration (https://github.com/Galooshi/import-js/issues/291)
2. Allow environments to specify full set of config (https://github.com/Galooshi/import-js/issues/290)
3. Move core modules into configuration (https://github.com/Galooshi/import-js/issues/292)
4. End global warming

(4 is a stretch goal, but would be nice to have). 😄

**Comment by rhettlivingston** [comment]
Will have to get to the rest of this later, but keyed in on this one...

> Interesting, I recently ran into a similar issue. I think ImportJS could probably also switch over to babel-eslint as well. Although I'd be interested to know if there's a performance difference between the two parsers.

Phooey. I thought you were already using babel-eslint, but now I see that it is just a devDependency. Hopefully, if there is a performance penalty it is minimal. Espree uses Acorn and Babylon evolved from Acorn. If you can't switch for everyone, we'll have to add a means to switch for users who need to. The Espree people just rejected our approach to create a pass-thru to Acorn (which supports our needs) today.

**Comment by rhettlivingston** [comment]
First, thank you, thank you, thank you for your time in all of this. Know that I plan on implementing all of the meteor specific stuff we're talking about here (plus things not spoken about that it enables) for sure and will also help on the rest.

This comment is geared toward the plan you started above.

Thought #1 - 

In looking at possible approaches to processing the configuration files, I looked at a few NPM modules specialized for that. One common feature was the ability to chain things. A system could start with defaults and add configurations from multiple configuration files before finally pulling them from the command line args last.

If in addressing #291 we implemented something like that, then we could pull the environments implementation out into canned configuration files that just got automatically processed when we encounter the environments statement. This would both ease maintenance and add to the understandability of the code.

Thought #2 - 

In looking at #230, it was interesting to me that some are putting this type of import at the top. We typically put it last.

This made me think about the whole groupedImports system. I think it could and should be generalized rather than trying to satisfy everyone with a fully canned system.

You have 3 groups defined now - core modules, package dependencies, and internal imports.

We're already talking about moving core module specification into the configuration as #292. Let's say we call that a "coreGroup" option that takes a list of modules. If someone also specifies an environment, it just adds to the list. While we're at it, we could add options for "npmPackageGroup" and "localGroup" (trying to think of a better term for internal imports). Both of these are automatically populated by your app already, but perhaps users have edge cases that they would like to add to them.

Then we could add a groupedImportsOrder option that would default to 
```json
   "groupedImportsOrder": ["coreGroup", "npmPackageGroup", "localGroup"]
```
The next step would be to allow new groups to be defined in the configuration file. Then I could define a meteorPackageGroup within the configuration system. 
```js
  "addGroups": ["meteorPackageGroup"],
  "meteorPackageGroup": [ ...  ],
```
I'm sure others have other groups to define as well. A bit of polish that would be needed in this step is to add some way to allow groups to be grouped. I'd want the npmPackageGroup and meteorPackageGroup to be combined. Maybe something like 
```json
   "groupedImportsOrder": [
      "coreGroup", 
      ["npmPackageGroup", "meteorPackageGroup"], 
      "localGroup"
   ]
```

Then, I could implement the identification and insertion of meteor packages into the packages section within the configuration system instead of in your app. You'd be much more future proof.

Finally, those of us with the side-effect imports could shift that group (however we go about defining it) to the top or bottom of the order or even combine it into other groups as they desire.

Thought #3 -
I need #230 fixed. 

Then I'll need something else to implement my side-effect imports. Want me to create the issue?

One thought is that with js configuration, perhaps a hook could be placed in a configuration file. You could pass it the file being processed and it could pass back imports that are required for that file that you wouldn't be able to identify from undefined vars. Another means of future proofing.

Thought #4 -

>I think we will have to add another exception to Configuration.resolveNamedExports (before we relativize the path).

Yep, want me to create the issue?

Thought #5 - 

If we switch to babel-eslint, we could just turn on allowImportExportEverywhere all of the time. It would have no detrimental side-effects to those not using nested imports. The parser just becomes more tolerant. Even if we don't outright switch, the only option we'd need to provide to the user is an option to utilize the babel-eslint parser instead of espree. There probably wouldn't be any reason not to enable every babel-eslint option.

Do you want me to create a "tolerate imports and exports below top-level" issue?

----

So,,, the full plan might be something like

1. Allow Meteor environments entry (#289) (mostly because a minimal shot is already done)
2. a first shot at Allow environments to specify full set of config (#290) (this is an easy addition to #289) (it will be added to as other things complete below)
3. add Configuration.resolveNamedExports exception for meteor packages (#???)
4. tolerate imports and exports below top-level (#???)
5. Allow javascript configuration (#291)
6. Move core modules into configuration (#292) (and implement other grouping improvements talked about above at same time).
7. Reimplement environments using configurations (#???)
8. Fix side-effect import issue in #230
9. Solve meteor's need for autoImport of components (.html and .css) (#???)
10. End global warming

(funny how that particular stretch goal keeps getting put off further and further) :smile:

**Comment by lencioni** [comment]
Here are some of my misc. notes:

> we could pull the environments implementation out into canned configuration files that just got automatically processed when we encounter the environments statement

This makes sense to me.

> In looking at #230, it was interesting to me that some are putting this type of import at the top. We typically put it last.

The reason for this is that this type of import by definition has side-effects. Since other modules might depend on those side-effects, the safest place to put it is at the top.

```json
  "addGroups": ["meteorPackageGroup"],
  "meteorPackageGroup": [ ...  ],
```

I'm not sure I fully understand this schema. What does `addGroups` do? What goes in the `meteorPackageGroup` array?

>  I'd want the npmPackageGroup and meteorPackageGroup to be combined.

Should combined groups get sorted together--essentially acting as one group all mixed together?

> Then I'll need something else to implement my side-effect imports. Want me to create the issue?

Yes, please create any new issues so we can keep discussions focused. I'd be happy to group them into a milestone so we can easily see the whole picture and track progress.

> Another means of future proofing.

Although I think a lot of this makes sense to me, I worry about over-future proofing things so that they become too complicated and over-engineered. I'm not sure we are doing that here, but I'd like us all to be careful that we only build things that we actually need and that are actually useful.

Thanks for collaborating on this with us!

**Comment by rhettlivingston** [comment]
>>In looking at #230, it was interesting to me that some are putting this type of import at the top. We typically put it last.
>The reason for this is that this type of import by definition has side-effects. Since other modules might depend on those side-effects, the safest place to put it is at the top.

Understood. But also something we don't have to worry about except in creating a default. My side-effects can go at the bottom. Other people may be interested in putting them at the top and the top sounds like a reasonable default for a lot of side-effect uses. 

With the availability of a groupedImportsOrder we could default to something like

```
   "groupedImportsOrder": ["sideEffectGroup", "coreGroup", "npmPackageGroup", "localGroup"],
```

and I could easily extend it for Meteor to something like

```
   "groupedImportsOrder": [
      "sideEffectGroup",
      "coreGroup", 
      ["npmPackageGroup", "meteorPackageGroup"], 
      "localGroup",
      "componentGroup"
   ]
```

where "componentGroup" contains my special side-effects.

>  ```json
>"addGroups": ["meteorPackageGroup"],
>  "meteorPackageGroup": [ ...  ],
> ```
>I'm not sure I fully understand this schema. What does addGroups do? What goes in the meteorPackageGroup array?

The concept here is to avoid having to think of every group a user, project, or environment might want. "addGroups" allows the user to define new groups. "meteorPackageGroup" in this case would insert meteor packages into the user-defined "meteorPackageGroup". Of course, this would actually be done by a function in one of the new js configuration files, not in a json file. I just put it in json for brainstorming.

>>I'd want the npmPackageGroup and meteorPackageGroup to be combined.
>Should combined groups get sorted together--essentially acting as one group all mixed together?

Exactly! I recognized that some might want a combined core and package group or just a core versus everything else separation. This simple adjustment allows users to combine their own groups with others including autogenerated ones in whatever way they desire.

>Yes, please create any new issues so we can keep discussions focused. I'd be happy to group them into a milestone so we can easily see the whole picture and track progress.

Thanks! Will get to it shortly.

>Although I think a lot of this makes sense to me, I worry about over-future proofing things so that they become too complicated and over-engineered. I'm not sure we are doing that here, but I'd like us all to be careful that we only build things that we actually need and that are actually useful.

I very much agree. An absolute must is to avoid disturbing the existing options. At the same time, the new advanced flexibility should be as clean and simple as possible to enable painless growth.

---

One other thing that's been on my mind, I lost a lot of time while first using this due to weird side-effects of errors I made in my .importjs.json file and didn't immediately track to there... repeatedly. I hope that in creating a configuration via javascript capability we can make it more robust in the validation and error handling and that error info gets out to the plugin users as something other than a strange plugin crash. I believe there is already an "enhancement" issue on that.

**Comment by lencioni** [comment]
> "addGroups" allows the user to define new groups.

Instead of having separate options, what do you think about having an `additionalGroups` option that is an object. Something like this:

```json
"additionalGroups": {
  "meteorPackageGroup": [ ... ],
}
```

**Comment by rhettlivingston** [comment]
>Instead of having separate options, what do you think about having an additionalGroups option that is an object. Something like this:
>
>```json
>"additionalGroups": {
>  "meteorPackageGroup": [ ... ],
>}
>```

What led me to the separate approach is that I'd still like to enable people to add to existing groups including coreGroup. So, for example, if I left something out of my Meteor coreGroup definition that "environments": ["meteor"] automagically created, a user could simply add another specification that would get merged with the one generated by the environment by saying (for example)

```json
"coreGroup": [ "meteor/jquery", "meteor/underscore" ],
```

This implied to me that defining a group and populating it are separate actions where the population step can be repeated in an additive fashion.

But, it's just brainstorming. I can imagine your approach working if I merge the whole structure with the whole existing internal structure each time I see it. In that case, I don't think the key name "additionalGroups" is correct... more like "groupAdditions" perhaps? The documentation could make clear then that adding to a group that doesn't exist actually creates it.

```json
"groupAdditions": {
  "coreGroup": [ "meteor/jquery", "meteor/underscore" ],
  "usersSpecialGroup": [ ... ],
}
```
A concern then is that we wouldn't be able to create a good error message if someone attempting to add to an existing group had a typo in the group name. We'd just create a new group with the typo'ed name. With the approach that I had, we'd be able to output a validation error on an unrecognized field in the configuration if they had not explicitly defined the group previously.

**Comment by lencioni** [comment]
Ah I see. I think `groupAdditions` makes sense. I'm a little uneasy of simply allowing arbitrary groups to be added to the top-level of the configuration. We could keep a separate `addGroups` option, but I'm not convinced the error message is worth it. I don't feel strongly either way.

**Comment by rhettlivingston** [comment]
Created
  #293 - Fix Meteor package imports when useRelativePaths is true
  #294 - Tolerate imports/exports below top-level
  #295 - Automatically recognize and import Meteor component files
  #296 - Reimplement environments using configurations

The revised plan now looks like...

1. Allow Meteor environments entry (#289) (mostly because a minimal shot is already done)
2. a first shot at Allow environments to specify full set of config (#290) (this is an easy addition to #289) (it will be added to as other things complete below)
3. add Configuration.resolveNamedExports exception for meteor packages (#293)
4. tolerate imports and exports below top-level (#294)
5. Allow javascript configuration (#291)
6. Move core modules into configuration (#292) (and implement other grouping improvements talked about above at same time).
7. Reimplement environments using configurations (#296)
8. Fix side-effect import issue in #230
9. Automatically recognize and import Meteor component files (#295)
10. End global warming

I'm planning to get PRs in place for review for the first four over the next few days.

**Comment by trotzig** [comment]
> I'm planning to get PRs in place for review for the first four over the next few days.

Fantastic! I'll try to be quick to review any code that you push. I'm usually checking github notifications at 9pm CET every day, and I'll try to check them a little more often now. I'm mostly afk for another few weeks though as I'm building a house. 

I can try to look at allowing javascript configuration in the coming few days. I also need to wrap up the work I've done on the [`faster-find` branch](https://github.com/Galooshi/import-js/tree/faster-find) before I start with other things. That branch is ~40 commits ahead of master, but I don't think there's anything in there that should affect the work we've planned out here much.

As you've already discussed, I'm a little concerned about the complexity of grouping. Going back to straight alphabetic ordering (no groups) is tempting, although I recognize that grouping has been important for some organizations to adopt import-js as a tool. Would it be terrible to disable grouping for the meteor environment to begin with until we have time to enhance it? 

Thanks for collaborating on this @rhettlivingston and @lencioni!

**Comment by rhettlivingston** [comment]
> I'm mostly afk for another few weeks though as I'm building a house.

Totally understand. I have about 3 months of house repairs on the back burner myself.

>As you've already discussed, I'm a little concerned about the complexity of grouping. Going back to straight alphabetic ordering (no groups) is tempting, although I recognize that grouping has been important for some organizations to adopt import-js as a tool. Would it be terrible to disable grouping for the meteor environment to begin with until we have time to enhance it?

As someone who implemented this type of automatic grouping of imports all the way back in the early '90s for other languages in large-scale development projects, I can attest to the craziness that can go into these coding standards in organizations.

Currently, without this and other changes we've discussed, the imports do come out as one big sorted group in Meteor. So, no need to disable. That's where we are at. Anything we do will just improve on that situation.

I think we can get a clean and flexible implementation of the grouping with minimal change. In reviewing your code, it mostly seemed to be a matter of generalizing what is already there. Moreover, I think I could work most of its aspects in parallel with javascript configurations. The js configuration capability merely expands its usefulness (an order of magnitude perhaps).

Anyway, no harm in a strawman. If we're not happy with it, I imagine it will be easy to adjust it till we are.

**Comment by rhettlivingston** [comment]
The original issue spawning this thread was resolved by 526c3d0 which was merged to master by 629e49b. Other issues have been created to address thoughts in the above dialog.

Closing this issue.
